libsweep/inc/protocol.h

```diff
@@ -111,18 +100,103 @@ typedef struct {
   uint8_t term;
 } SWEEP_PACKED response_info_version_s;
 
-static_assert(sizeof(response_info_version_s) == 21, "response info version size mismatch");
-
 typedef struct {
   uint8_t cmdByte1;
   uint8_t cmdByte2;
   uint8_t motor_speed[2];
   uint8_t term;
 } SWEEP_PACKED response_info_motor_s;
+#else
+#if defined _WIN32 || defined __CYGWIN__
+SWEEP_PACKED(typedef struct {
+  uint8_t cmdByte1;
+  uint8_t cmdByte2;
+  uint8_t cmdParamTerm;
+} cmd_packet_s;)
 
-static_assert(sizeof(response_info_motor_s) == 5, "response info motor size mismatch");
+SWEEP_PACKED(typedef struct {
+  uint8_t cmdByte1;
+  uint8_t cmdByte2;
+  uint8_t cmdParamByte1;
+  uint8_t cmdParamByte2;
+  uint8_t cmdParamTerm;
+} cmd_param_packet_s;)
+
+SWEEP_PACKED(typedef struct {
+  uint8_t cmdByte1;
+  uint8_t cmdByte2;
+  uint8_t cmdStatusByte1;
+  uint8_t cmdStatusByte2;
+  uint8_t cmdSum;
+  uint8_t term1;
+} response_header_s;)
+
+SWEEP_PACKED(typedef struct {
+  uint8_t cmdByte1;
+  uint8_t cmdByte2;
+  uint8_t cmdParamByte1;
+  uint8_t cmdParamByte2;
+  uint8_t term1;
+  uint8_t cmdStatusByte1;
+  uint8_t cmdStatusByte2;
+  uint8_t cmdSum;
+  uint8_t term2;
+} response_param_s;)
+
+SWEEP_PACKED(typedef struct {
+  uint8_t sync_error;
+  uint16_t angle; // see u16_to_f32
+  uint16_t distance;
+  uint8_t signal_strength;
+  uint8_t checksum;
+} response_scan_packet_s;)
 
-// Read and write specific packets
+SWEEP_PACKED(typedef struct {
+  uint8_t cmdByte1;
+  uint8_t cmdByte2;
+  uint8_t bit_rate[6];
+  uint8_t laser_state;
+  uint8_t mode;
+  uint8_t diagnostic;
+  uint8_t motor_speed[2];
+  uint8_t sample_rate[4];
+  uint8_t term;
+} response_info_device_s;)
+
+SWEEP_PACKED(typedef struct {
+  uint8_t cmdByte1;
+  uint8_t cmdByte2;
+  uint8_t model[5];
+  uint8_t protocol_major;
+  uint8_t protocol_min;
+  uint8_t firmware_major;
+  uint8_t firmware_minor;
+  uint8_t hardware_version;
+  uint8_t serial_no[8];
+  uint8_t term;
+} response_info_version_s;)
+
+SWEEP_PACKED(typedef struct {
+  uint8_t cmdByte1;
+  uint8_t cmdByte2;
+  uint8_t motor_speed[2];
+  uint8_t term;
+} response_info_motor_s;)
+#else
+#error "Only Clang, GCC and Visual supported at the moment, please open a ticket"
+#define SWEEP_API
+#define SWEEP_PACKED
+#endif
+#endif
```

Comment:
Hm this got super ugly - we need to figure out a better way to handle these struct / packed issues..